source/idf/EthernetDevice.cpp

```diff
@@ -0,0 +1,136 @@
+    sockType(SOCK_STREAM),
+    tcp(true) {
+        memset(&serverAddr, 0 , sizeof(serverAddr));
+        setTCP(); // ensure vars are zeroed out.
+    }
+
+void EthernetDevice::open() {
+    if (!mOpen) {
+
+        std::ostringstream stream;
+        stream << "[IDF::EthernetDevice::open()] ";
+
+        socketHandle = socket(AF_INET, sockType, 0);
+
+        if ( errno > 0) {
+            stream << "failed to create socket";
+            perror(stream.str().c_str());
+            throw IOException(stream.str());
+        }
+
+        serverAddr.sin_family = AF_INET;
+        serverAddr.sin_port = htons(serverPort);
+        serverAddr.sin_addr.s_addr = inet_addr(serverName.c_str());
+        serverAddrLen = sizeof(serverAddr);
+
+        std::cout << "[IDF::EthernetDevice] Connecting to " << serverName << ":" << serverPort << std::endl;
+
+        char sendBuffer[] = "hello HC";
+        int ret = -1;
+        while (1) {
+            ret = tcp ? connect(socketHandle, (struct sockaddr*)&serverAddr, serverAddrLen)
+                      : sendto(socketHandle, sendBuffer, sizeof(sendBuffer), 0, (struct sockaddr *)&serverAddr, serverAddrLen);
+            if(ret < 0) {
+                if (errno == EINTR) { continue; } // interrupted by a SIGNAL; retry
+                stream << "failed to connect to " << (tcp ? "TCP" : "UDP") << " device " << serverName << ":" << serverPort;
+                perror(stream.str().c_str());
+                throw IOException(stream.str());
+            }
+            break;
+        }
+
+        std::cout << "[IDF::EthernetDevice] Connected to " << serverName << ":" << serverPort << std::endl;
+
+        Manageable::open();
+    }
+}
+
+/** disconnects from the server */
+void EthernetDevice::close() {
+    if (mOpen) {
+        ::close(socketHandle);
+        Manageable::close();
+    }
+}
+
+std::vector<std::vector<unsigned char> > EthernetDevice::read() {
+    std::vector<std::vector<unsigned char> > results;
+    std::vector<unsigned char> buffer(packetLength);
+
+    while (read(&buffer[0], buffer.size())) {
+        results.push_back(buffer);
+    }
+
+    return results;
+}
+
+unsigned EthernetDevice::read(unsigned char *buffer, size_t length) {
+    if (!mOpen) {
+        open();
+    }
+
+    int bytesRecvd = 0;
+    unsigned bytesTotal = 0;
+    unsigned char readBuff[length*10];
+    memset(&readBuff, 0, length);
+
+    while(bytesTotal < length) {
+        // non-blocking receive
+        bytesRecvd = recvfrom(socketHandle, &readBuff[bytesTotal], length-bytesTotal, MSG_DONTWAIT, (struct sockaddr*)&srcAddr, &srcAddrLen);
+        if (bytesRecvd < 0) {
+            if (errno == EAGAIN) { return 0; } // no data, try again later
+            else if (errno == EINTR) { continue; } // interrupted, retry
+            else {
+                close();
+                std::ostringstream stream;
+                stream << "Error while reading from " << name;
+                perror(stream.str().c_str());
```

Comment:
`EthernetDevice::write` throws an `IOException` on error. Is there a reason `read`'s behavior is different?

Reply:
haven't paid serious attention to write yet. that's one of the reasons this is still a draft. So far this was developed with just a EthernetDevice version of the Extreme3dPro, which doesn't have any need for writes. I'm now working on the interface for the end device, which we cannot release publicly.

Plan to come back here and update once that gets to the stage of writing.

Reply:
> `EthernetDevice::write` throws an `IOException` on error. Is there a reason `read`'s behavior is different?

follow up question: is IOException the preferred IDF idiom?

Reply:
Yes. Users cannot programmatically respond to a print out, but they can catch an exception and decide what do about it.